config/pkcs11/config.go

```diff
@@ -0,0 +1,116 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package pkcs11config
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/containers/ocicrypt/crypto/pkcs11"
+	enchelpers "github.com/containers/ocicrypt/helpers"
+	"gopkg.in/yaml.v2"
+)
+
+// OcicryptoConfig represents the format of an imgcrypt.conf config file
+type OcicryptConfig struct {
+	Pkcs11Config pkcs11.Pkcs11Config `yaml:"pkcs11"`
+}
+
+const CONFIGFILE = "ocicrypt-pkcs11.conf"
+
+// parseConfigFile parses a configuration file; it is not an error if the configuration file does
+// not exist, so no error is returned.
+func parseConfigFile(filename string) (*OcicryptConfig, error) {
+	// a non-existent config file is not an error
+	_, err := os.Stat(filename)
+	if os.IsNotExist(err) {
+		return nil, nil
+	}
+
+	data, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	ic := &OcicryptConfig{}
+	err = yaml.Unmarshal(data, ic)
+	return ic, err
+}
+
+// getConfiguration tries to read the configuration file at the following locations
+// 1) ${OCICRYPT_PKCS11_CONFIG}
+// 2) ${XDG_CONFIG_HOME}/ocicrypt-pkcs11.conf
+// 3) ${HOME}/.config/ocicrypt-pkcs11.conf
+// 4) /etc/ocicrypt-pkcs11.conf
+// If no configuration file could be found or read a null pointer is returned
+func getConfiguration() (*OcicryptConfig, error) {
+	filename := os.Getenv("OCICRYPT_PKCS11_CONFIG")
+	if len(filename) > 0 {
+		ic, err := parseConfigFile(filename)
+		if err != nil || ic != nil {
+			return ic, nil
```

Comment:
it seems like there's some inconsistencies with err returning. if `err != nil` should the error be returned here? or should there be a fallthrough to the next case, it seems like the only place where an err can be returned is in the last `/etc/` case (and errors are suppressed for other cases).

Reply:
it should be `return ic,nil` in all these cases here. Will fix.